core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/worker/DataProcessor.scala

```diff
@@ -93,12 +93,47 @@ class DataProcessor(
     }
   }
 
+  private[this] def processInputState(state: State, port: Int): Unit = {
+    try {
+      val outputState = executor.processState(state, port)
+      if (outputState.isDefined) {
+        outputManager.emitMarker(outputState.get)
+      }
+    } catch safely {
+      case e =>
+        // forward input tuple to the user and pause DP thread
+        handleExecutorException(e)
+    }
+  }
+
+  /**
+    * process start of an input port with Executor.onStart().
+    * this function is only called by the DP thread.
+    */
+  private[this] def processStartOfUpstream(portId: Int): Unit = {
+    try {
+      outputManager.emitMarker(StartOfUpstream())
+      val outputState = executor.produceStateOnStart(portId)
+      if (outputState.isDefined) {
+        outputManager.emitMarker(outputState.get)
+      }
+    } catch safely {
+      case e =>
+        // forward input tuple to the user and pause DP thread
+        handleExecutorException(e)
+    }
+  }
+
   /**
     * process end of an input port with Executor.onFinish().
     * this function is only called by the DP thread.
     */
   private[this] def processEndOfUpstream(portId: Int): Unit = {
     try {
+      val outputState = executor.produceStateOnFinish(portId)
+      if (outputState != null && outputState.isDefined) {
```

Comment:
will it return null?

Reply:
It will return null in the DP test case. I will try to fix the test case.

Reply:
Fixed.